algviz/picture/connection_map.py

```diff
@@ -0,0 +1,127 @@
+            self.start_point = start_point
+            self.connect_pic = connect_pic
+            if not connect_pic.is_drawn:
+                connect_pic.draw()
+            self.anchor = anchor
+            self.left_shift_y = 0
+            self.right_shift_y = 0
+
+        def is_pointer(self):
+            return False
+
+    class PointerConnection(Connection):
+        def __init__(self, start_point, connect_pic, anchor, pointer_style={}):
+            self.pointer_style = pointer_style
+            super().__init__(self, start_point, connect_pic, anchor)
+
+        def is_pointer(self):
+            return True
+    
+    def __init__(self, startPicture):
+        self.connections = []
+        self.picture = startPicture
+        
+    def add_connection(self, point, connect_pic, anchor=None, pointer=True, pointer_style={}):
+        if self.picture.width < point[0] or self.picture.height < point[1]:
+            raise TypeError("Point not included in picture")
+        if pointer:
+            self.connections.append(ConnectionMap.PointerConnection(point, connect_pic, anchor, pointer_style))
+        else:
+            self.connections.append(ConnectionMap.Connection(point, connect_pic, anchor))
+
+    def draw_connections(self):
+        #TODO: reimplement with GraphViz?
+        cur_pic_width = self.picture.width
+        cur_pic_height = self.picture.height
+        left_width = 0
+        right_width = 0
+        left_height = 0
+        right_height = 0
+        for connection in self.connections:
+            if not connection.is_pointer():
+                #in place connections dont change size of picture
+                continue
+            connect_size = connection.connect_pic.size
+            if connection.anchor == Anchor.LEFT:
+                #left anchor = picture should be placed to right of main picture
+                right_width = max(connect_size[0], right_width)
+                right_height += connect_size[1]
+            elif connection.anchor == Anchor.RIGHT:
+                #right anchor = picture placed on left of main picture
+                left_width = max(connection_size[1], left_width)
+                left_height += connect_size[1]
+            else:
+                #Otherwise, picture connected on whatever side its pointer is closes to
+                #There are better ways to do this, but I wont improve unless graphviz doesn't pan out
+                if connection.start_point[0] <= self.picture_size[0]/2:
+                    left_width = max(connection_size[1], left_width)
+                    left_height += connect_size[1]
+                else:
+                    right_width = max(connect_size[0], right_width)
+                    right_height += connect_size[1]
+        new_width = left_width + cur_pic_width + right_width
+        new_height = max(left_height, cur_pic_height, right_height)
+        main_svg = svgutils.fromfile(self.picture.filename)
```

Comment:
We desperately need to encapsulate the fact that every picture is a file with a filename.  We don't even need to fix it -- just hide it so that we forget about it until later we wonder "why is algviz taking so long to generate pictures?"  And then we'll fix it.